File: app/web/src/api/sdf/dal/remote_shell.ts
import { sdfApiInstance } from "@/store/apis.web";

export interface CreateRemoteShellSessionRequest {
  image?: string;
  workingDir?: string;
  envVars?: Record<string, string>;
}

export interface RemoteShellConnectionInfo {
  natsSubject: string;
  stdinSubject: string;
  stdoutSubject: string;
  stderrSubject: string;
  controlSubject: string;
}

export enum RemoteShellStatus {
  Active = "active",
  Error = "error", 
  Terminated = "terminated",
}

export interface CreateRemoteShellSessionResponse {
  executionId: string;
  sessionId: string;
  containerId: string;
  connectionInfo: RemoteShellConnectionInfo;
  status: RemoteShellStatus;
  message?: string;
}

export interface RemoteShellSessionDetails {
  forcedChangeSetId: string;
  data: CreateRemoteShellSessionResponse;
}

export class RemoteShellApi {
  static async createSession(
    workspaceId: string,
    changeSetId: string,
    request: CreateRemoteShellSessionRequest = {}
  ): Promise<RemoteShellSessionDetails> {
    const response = await sdfApiInstance.post(
      `/v2/workspaces/${workspaceId}/change-sets/${changeSetId}/remote-shell/create`,
      request
    );

    // The backend returns a ForceChangeSetResponse where:
    // - The actual data is in response.data directly (CreateRemoteShellSessionResponse)
    // - The force_change_set_id is in the response headers
    return {
      forcedChangeSetId: response.headers.force_change_set_id || '',
      data: response.data as CreateRemoteShellSessionResponse,
    };
  }
}

File: lib/cyclone-core/src/remote_shell.rs
use serde::{
    Deserialize,
    Serialize,
};
use telemetry::prelude::*;
use telemetry_utils::metric;

use crate::CycloneRequestable;

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RemoteShellRequest {
    pub execution_id: String,
    pub image: Option<String>,
    pub env_vars: std::collections::HashMap<String, String>,
    pub working_dir: Option<String>,
}

#[derive(Debug, Deserialize, Eq, PartialEq, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct RemoteShellResultSuccess {
    pub execution_id: String,
    pub session_id: String,
    pub container_id: String,
    pub connection_info: RemoteShellConnectionInfo,
    pub status: RemoteShellStatus,
    pub message: Option<String>,
}

#[derive(Debug, Deserialize, Eq, PartialEq, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct RemoteShellConnectionInfo {
    pub nats_subject: String,
    pub stdin_subject: String,
    pub stdout_subject: String,
    pub stderr_subject: String,
    pub control_subject: String,
}

#[remain::sorted]
#[derive(Debug, Deserialize, Eq, PartialEq, Serialize, Clone, Copy)]
#[serde(rename_all = "camelCase")]
pub enum RemoteShellStatus {
    Active,
    Error,
    Terminated,
}

impl CycloneRequestable for RemoteShellRequest {
    type Response = RemoteShellResultSuccess;

    fn execution_id(&self) -> &str {
        &self.execution_id
    }

    fn kind(&self) -> &str {
        "remoteShell"
    }

    fn websocket_path(&self) -> &str {
        "/execute/remote-shell"
    }

    fn inc_run_metric(&self) {
        metric!(counter.function_run.remote_shell = 1);
    }

    fn dec_run_metric(&self) {
        metric!(counter.function_run.remote_shell = -1);
    }
}